### DISCUSSION TITLE
Custom in-memory cache with NodeCache | Next.Js 14

### DISCUSSION THREAD

**Discussion by smo043** [vercel/next.js#59509] [Help] [answered]
I am looking for an example of how to use node-cache to implement an in-memory custom cache, because the custom in-memory cache is removed with each API request. I tried several methods, but none of them worked. I am using the latest Next.Js 14.

Here's what I have tried:

I created an api route with NodeCache initialization. When the api is called for the first time after signIn, we save some information in memory because the same data will be used for the next 30 minutes. 

When I call the api again after the data is saved for the first time, the NodeCache is empty and it hits the actual api, which I want to avoid. Any assistance would be greatly appreciated.

**Comment by eduardodallmann** [comment, marked as answer]
Hello @smo043, one way to do this is to use instrumentation. [docs](https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation)

1. First, create the `instrumentation.ts` file in the root of `src` with this code:
```
import type NodeCache from 'node-cache';

export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    const NodeCache = (await import('node-cache')).default;
    const config: NodeCache.Options = {
      stdTTL: process.env.NODE_ENV === 'production' ? 0 : 60,
    };

    global.cacheConfigs = new NodeCache(config);
    global.cacheUser = new NodeCache(config);
    //other caches

    // Other things that should only happen once at server startup.
    // firebase-admin connect, etc
  }
}
```
2. Add this to `next.config`:
```
  experimental: {
    instrumentationHook: true,
  },
```
3. For typescript only: Create a `global.d.ts` file in the project root with this:
```
import type NodeCache from 'node-cache';

declare global {
  var cacheConfigs: NodeCache;
  var cacheUser: NodeCache;
  //your cache names
}

export {};
```
_PS: You may need to include `global.d.ts` in tsconfig depending on how it is configured._


4. Ok, now the cache will not be reset with each request, only at server startup. Whenever you need to access the cache in a serve component, server action, etc., use global first. Like this:
```
global.cacheUser.set('foo', 'bla');
```

I hope this helps.

**Reply by smo043** [reply]
@eduardodallmann - Thank you, amazing, let me try this!

**Comment by Vacilando** [comment]
Have you had success implementing this, @smo043 - any code or additional insights you can share?

**Reply by smo043** [reply]
I simply used the same snippet provided in the answer.

**Reply by invisibleved** [reply]
@smo043 I have a problem with the product mode. do you have this problem?
